### PR TITLE
Parser: Remove specific support for <!--more--> tag

### LINKED ISSUE #2973: Parser: Move support for the <!--more--> tag out of the parser
`<!--more-->` (and its cousin, `<!--noteaser-->`) are currently handled as special cases in the parser.

To keep the parser Gutenberg-focussed, these should be moved to the legacy paste handler, where they can be transformed into a `<!-- wp:more -->` block.

### PR DESCRIPTION
**Pull request by mcsf**
## Description
Fixes #2973. This pull request:
- removes the exceptions to the block parser (`<!--more-->` and `<!--teaser-->`), thus keeping the parser "Gutenberg-focus", as @pento put it;
- treats the More block _more or less_ as any other block, meaning it surrounds the `more` and optional `noteaser` tags with regular comment demarcations;
- retains backwards compatibilty through regular block conversion mechanisms (from freeform via "Convert to Blocks");
- paves the way for implementing a Pagination (`<!--nextpage-->`; see #4930) block without introducing further parser exceptions.

## How Has This Been Tested?

Pasting hasn't yet been tested. For conversion of legacy content, try:
1. Using the classic editor, build a post (a lorem ipsum with six paragraphs should suffice). Sprinkle it with the following:

```
<!--more-->
```

and

```
<!--more-->
<!--noteaser-->
```

and

```
<!--more Read all about it!-->
<!--noteaser-->
```

2. Save it as a draft.
3. Open the post in Gutenberg.
4. Find the freeform (Classic) blocks corresponding to the `more` tags.
5. Convert them to blocks (_advanced block settings > convert to blocks_).
6. Make sure the new More blocks feature the right attributes (`customText` and `noTeaser`). The `customText` attribute can be observed by looking at the text on the More block's visual representation; the `noTeaser` attribute can be observed with the inspector controls.

### Caveat

- `rawHandler`'s transformers will get rid of HTML comments, and will attempt to merge or eliminate many elements (e.g. paragraphs, line breaks). In order to get around this without making special exceptions for certain comments, `commentRemover` will detect `more` and `noteaser` comments and replace them with a custom element `<wp-block>`. This element has no particular definition, is never rendered, and is solely used in order to pass block information through `rawHandler`. The More block's `raw` transform will then match against its shape to replace it with a proper block.
- `commentRemover` happened to be the place where these operations were initially introduced, but the transformer can be split in two.
- The unit tests for the serializer were kept and amended but, as explained in a comment, they should probably be removed. Right now, they illustrate the overall changes of this PR.

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style.
- [ ] My code has proper inline documentation.

### CODE REVIEW

File: blocks/api/raw-handling/comment-remover.js
Comment (dmsnell) [inline blocks/api/raw-handling/comment-remover.js]: this RegExp is pretty much asking for trouble. could we whitelist characters instead of blacklisting them?

```
<!--moredanghtmlcomments-->
```

**Comment by dmsnell** [comment]
Love this! Thank you!

File: blocks/api/raw-handling/comment-remover.js
Comment (mcsf) [reply blocks/api/raw-handling/comment-remover.js]: Thanks for having a look. I'm not sure what you mean by whitelisting, though. FWIW, this is how core does it:

https://github.com/WordPress/WordPress/blob/e50b9c943fddf781f980bba78d1ce927a5dc11b4/wp-includes/post.php#L532-L534

Comment (dmsnell) [reply blocks/api/raw-handling/comment-remover.js]: Is this a valid more tag?

```
<!--moreYou should continue reading below!-->
```

Comment (mcsf) [reply blocks/api/raw-handling/comment-remover.js]: It is valid. :)

screen shot 2018-02-15 at 18 17 25

Agreed on the greed, though.

Comment (dmsnell) [reply blocks/api/raw-handling/comment-remover.js]: also, `/more(.*)/` is redundant. it matches the same things that `/more/` matches because both _nothing_ and _everything_ is `.*`

Comment (mcsf) [reply blocks/api/raw-handling/comment-remover.js]: > zero of everything and one of everything and many of everything is `.*`

philosoraptor

>  `/more(.*)/` is redundant

Fair enough. We already check for `indexOf( 'more' ) === 0`, so why don't we just grab the substring starting after `'more'`? Then we can `.trim()` it. No RegExp needed.

Comment (dmsnell) [reply blocks/api/raw-handling/comment-remover.js]: `nodeValue.slice(4).trim()`

**Comment by mcsf** [comment]
Exactly.

File: blocks/library/more/index.js
Comment (youknowriad) [inline blocks/library/more/index.js]: So the `dataset` is used here and it's not a generic API right? it's just a specific attribute created by the `special-comment-converter.js` which is also specific to this bloc.

This makes me wonder, if this converter's code shouldn't be written somewhere in this transform instead?

File: blocks/api/raw-handling/special-comment-converter.js
Comment (youknowriad) [inline blocks/api/raw-handling/special-comment-converter.js]: Do you think we should unit test this or is it already covered elsewhere?

**Review by youknowriad — APPROVED** [review]
LGTM 👍 This surfaces this though https://github.com/WordPress/gutenberg/issues/4958

File: blocks/api/raw-handling/special-comment-converter.js
Comment (mcsf) [reply blocks/api/raw-handling/special-comment-converter.js]: I think some tests are in order 👍

File: blocks/library/more/index.js
Comment (mcsf) [reply blocks/library/more/index.js]: > it's not a generic API right? it's just a specific attribute created by the special-comment-converter.js which is also specific to this bloc.

Correct. It is a bit of a fake separation. However, I named it `special-comment-converter` instead of `more-converter` because I'm anticipating its use for `<!--nextpage-->` later.

> This makes me wonder if this converter's code shouldn't be written somewhere in this transform instead?

Hm. Maybe I looked at it the wrong way, but basically dataset + custom tag was what I came up with to _preserve_ the more tag and not let it get destroyed by the rest of `rawHandler`. This seems like a better trade-off than e.g. making exceptions in the other transformers to ignore `more`. What do you think? Also pinging @iseulde for ideas. :)

Comment (iseulde) [reply blocks/library/more/index.js]: I agree that the separation is weird, but can't think of anything better right now... Hm or maybe? The raw handler itself is supposed to return clean HTML which blocks can hook into to transform. We already have shortcodes in this mix too though, but the transform is `shortcode` instead of `raw`. Maybe we could follow the same pattern and have a `comment` transform? If no block transforms the comment, then we should just drop the comment (not pass through `raw` transforms). What do you think? This removes the need for the special case comment converter.

**Comment by iseulde** [comment]
Okay, After looking a bit at the code, The approach in https://github.com/WordPress/gutenberg/pull/5061#discussion_r169285395 might be simpler too? Also no `isMatch` function needed.

**Comment by mcsf** [comment]
> Maybe we could follow the same pattern and have a `comment` transform?

This made sense to me initially, but less so after giving more thought. The main complexity is that, unlike with `shortcode` transforms, we don't have a 1:1 relationship between comments and blocks, due to the way `more` compounds with `noteaser`. That is, a transform interface like:

```js
{ type: 'comment', /* isMatch, */, transform( commentNode ) { } }
```

isn't enough, since we need to look further than just the single node. Now, the `transform` _could_ be clever and look at `commentNode`'s siblings and selectively destroy them, but that doesn't seem like a sound approach—that is, I can expect functions within `rawHandler`'s `deepFilterHTML` calls to be destructive in breadth, but not the handlers of individual block types.

My complementary question is: how many more comments can we expect to have to parse in Gutenberg to justify the introduction of a new public transform type? Right now we can only expect WordPress legacy markings: `more`, `noteaser`, `nextpage`; ideally, new block types will only care about proper HTML (`raw` transforms) and shortcodes (`shortcode`). I'd say that, if an author wants to add support for a very specific input format that includes complex shapes with HTML comments, then they'll need something more powerful, like a hook to access the filter sets of `rawHandler`, further discarding the need for a new `comment`-type transform. Does this reasoning make sense, @iseulde?

File: blocks/api/raw-handling/special-comment-converter.js
Comment (pento) [inline blocks/api/raw-handling/special-comment-converter.js]: Amusing side note: this fixes a bug in the `more` handling currently in the parser, which requires a space between "more" and the custom text. Core doesn't require that space. 🙂

Comment (pento) [reply blocks/api/raw-handling/special-comment-converter.js]: Oh, I missed the previous discussion wherein @dmsnell learned yet another wonderful weird thing about Core. 🙃

**Review by pento — APPROVED** [review]
👍🏻 on this.

File: blocks/api/raw-handling/special-comment-converter.js
Comment (iseulde) [inline blocks/api/raw-handling/special-comment-converter.js]: Assuming that the intention is to ensure that the more element is at the top level (to convert into blocks later), could we loop through parent nodes until the body node is the parent of that node?

```js
while ( parent.nodeName !== 'BODY' ) {
     parent = parent.parentNode;
}
```

Also not completely sure what "appending to the closer P parents fails" means.

File: blocks/api/raw-handling/special-comment-converter.js
Comment (iseulde) [inline blocks/api/raw-handling/special-comment-converter.js]: Interesting thing :) I wonder if we can also do this for shortcakes... Might enable us to get rid of the "pieces" of HTML and parse it all in one go.

**Review by ellatrix — APPROVED** [review]

File: blocks/api/raw-handling/special-comment-converter.js
Comment (mcsf) [reply blocks/api/raw-handling/special-comment-converter.js]: > Assuming that the intention is to ensure that the more element is at the top level (to convert into blocks later), could we loop through parent nodes until the body node is the parent of that node?

This is more straightforward, I like it.

> Also not completely sure what "appending to the closer P parents fails" means.

The `<wp-block>` element couldn't be appended to any `<p>`; inspecting the parent would show that no children had been added.

**Comment by mcsf** [comment]
I'll wait until we have #4958 fixed before merging this one. Thanks for the reviews!

* * *

Worth pointing out that, in an impromptu discussion with @mtias, it was brought up that in the long run the `<!-- wp:more --/>` block shouldn't need to contain the legacy comments, as it itself would be the primitive that WordPress would understand for partitioning a post. However, core WP doesn't really offer the needed hooks to change that, so this would have to be handled in one of two ways:
1. Wait for the merge into core to change core itself.
2. Add Gutenberg-only support for `<!-- wp:more --/>` on the server—different approaches are possible, from adding a specific rule to handle that block (pseudo-block?), to getting creative with More's `render_callback` (which I'm not convinced is good).

I'm leaning towards number 1.

**Pull request merged**
